### PR TITLE
Dockerize Prefect Server, Agent and Proxy

### PR DESCRIPTION
**Pull request by d-jeph**
### Issue
In the current setup it's not possible to run prefect, prefect agent and the proxy as docker containers. This PR adds the necessary docker files to enable running these services as containers.

### Changes
 Added the following files and folders:
- A `Docker` folder with a `docker_compose.yml` file and `env.template` file. The `docker_compose` file has three services: `prefect_proxy`,`server`, and `agent`
- A `Dockerfile` for the proxy python application
- The `env.template` file contains the necessary environment variables for the `docker_compose` file
- A `docker_compose.dev.yml` for running the services locally during development and testing
- A `docker.yml` that contains a Github actions workflow for pushing the docker images to docker hub

### CODE REVIEW

**Comment by Ishankoradia** [comment]
Hi @d-jeph , thank you submitting this PR. Its super helpful. 

For prefect, we have upgraded from `agents` to `workers`.  This [issue](https://github.com/DalgoT4D/platform_infra/issues/35) has the details 
Can we incorporate this in the `docker-compose.yml` ?

Also will we be able to start mutliple workers in the current setup ? On production we use 3 - 2 workers execute work from `ddp` queue and 1 from `manual-dbt` queue.

**Comment by d-jeph** [comment]
> Hi @d-jeph , thank you submitting this PR. Its super helpful.
> 
> For prefect, we have upgraded from `agents` to `workers`. This [issue](https://github.com/DalgoT4D/platform_infra/issues/35) has the details Can we incorporate this in the `docker-compose.yml` ?
> 
> Also will we be able to start mutliple workers in the current setup ? On production we use 3 - 2 workers execute work from `ddp` queue and 1 from `manual-dbt` queue.

@Ishankoradia Thanks. I will update the code to use workers and also the multiple workers

**Comment by d-jeph** [comment]
> > Hi @d-jeph , thank you submitting this PR. Its super helpful.
> > For prefect, we have upgraded from `agents` to `workers`. This [issue](https://github.com/DalgoT4D/platform_infra/issues/35) has the details Can we incorporate this in the `docker-compose.yml` ?
> > Also will we be able to start mutliple workers in the current setup ? On production we use 3 - 2 workers execute work from `ddp` queue and 1 from `manual-dbt` queue.
> 
> @Ishankoradia Thanks. I will update the code to use workers and also the multiple workers

@Ishankoradia I have pushed the updated `docker-compose.yml`. I could not find a straight forward way for making the number of workers dynamic so for now I have three different services, one for each worker

**Comment by d-jeph** [comment]
@fatchat I added a github actions workflow change to this PR as well.

**Comment by Ishankoradia** [comment]
Thanks @d-jeph i will take a look at this and merge it before the weekend.

**Review by Ishankoradia — CHANGES_REQUESTED** [review]

File: Docker/docker-compose.dev.yml [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    image: prefecthq/prefect:2.13.5-python3.10
+    restart: always
+    depends_on:
+      - prefect_postgres_db
+    entrypoint: ["/bin/sh", "/opt/prefect/start_prefect_server.sh"]
+    environment:
+      - PREFECT_UI_URL=${PREFECT_API_URL}
+      - PREFECT_API_URL=${PREFECT_API_URL}
+      - PREFECT_SERVER_API_HOST=0.0.0.0
+      - PREFECT_API_DATABASE_CONNECTION_URL=${DB_CONNECTION_URL}
```
[the remark this answers is not in the document]
Comment (d-jeph) [reply Docker/docker-compose.dev.yml]: hardcoded the connection url using the db service credentials

File: Docker/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      - PREFECT_API_URL=${PREFECT_API_URL}
+    ports:
+      - 4300:4300
+
+  ### Prefect Server API and UI
+  prefect_server:
+    image: prefecthq/prefect:2.13.5-python3.10
+    restart: always
+    depends_on:
+      - prefect_postgres_db
```
[the remark this answers is not in the document]
Comment (d-jeph) [reply Docker/docker-compose.yml]: Thanks - I made the correct

File: Docker/docker-compose.dev.yml [hunk 3]
```diff
@@ -0,0 +1,106 @@
+version: "3.9"
+services:
+  ## Test DB
+  prefect_postgres_db:
+    container_name: prefect_db
+    image: postgres
+    environment:
+      POSTGRES_USER: ${POSTGRES_USER:-postgres}
+      POSTGRES_PASSWORD: ${POSTGRES_PASSWORD:-securepass}
+      PGDATA: /data/postgres
```
[the remark this answers is not in the document]
Comment (d-jeph) [reply Docker/docker-compose.dev.yml:11]: Postgres should now automatically create the db using the value provided in the `POSTGRES_DB` var

**Comment by Ishankoradia** [comment]
Hi @d-jeph , the docker compose works very nicely. I am able to run proxy + prefect on my current local database via docker.

One thing i noticed, when i run an existing prefect deployment (from its UI, where prefect is running via docker). It throws this error. Although the flow is getting submitted to the container worker but looks like it is still trying to find the code (to be run) from the host. 
Wanted to check if you have faced this before, looks we might have to update the existing deployment to read from prefect-proxy's storage instead of host somehow.

Screenshot 2024-06-21 at 18 05 17

**Comment by d-jeph** [comment]
@Ishankoradia Thanks for flagging this. 
Yes we experienced the same issue when we tested this on GCP. 
I have updated the docker compose files to add a shared volume between the proxy service(where the flow code lives) and the prefect worker containers. This should help prefect get access to the flow code. Please test and let me know if you get any errors.

**Comment by Ishankoradia** [comment]
Hey @d-jeph thanks for making the changes. This is what i found in testing experiment
- I ran `docker compose up` on my local prefect that has whole lot of deployments setup
- I tried running one of them , it gave the same error as above. To make it work i had to go into prefect's database, navigate to table `depolyment` and update the column `path` to `/app`.  This is absolutely fine, we could run a script that does this for our current deployments if/when we move to dockerize production deployment.
- After making the above change, i again ran the same deployment & this time i got the below error. (Looks like its finding the code now)
Screenshot 2024-06-25 at 20 51 46

**Comment by d-jeph** [comment]
@Ishankoradia Thanks for sharing this. 
I have now setup a new Docker file `Dockerfile.prefect_server` that creates a prefect server image that includes the same dependencies as the proxy service - this is the image used now in the docker compose for the prefect server and worker service. When I used the proxy `requirements.txt` I however got an error `cannot import name 'safe _get_annotation' from 'griffe.agents-nodes'` so I had to upgrade the `griffe` package to the newest version `0.47.0`. Please let me know if you've encountered the error

I have also noticed the team upgraded prefect to a new version `2.18.3` so I upgraded the image used as well.

Please test and let me know if you experience any errors on your end.

**Comment by Ishankoradia** [comment]
Hi @d-jeph , i have tested it out. Thanks for making the changes and ugprading the images with the newer prefect version.

Just want to reconfirm that this is still true ? (second point from my last comment)
```
I tried running one of them , it gave the same error as above. To make it work i had to go into prefect's database, navigate to table depolyment and update the column path to /app. This is absolutely fine, we could run a script that does this for our current deployments if/when we move to dockerize production deployment.
```

After doing the above, it is able to download the code & submit flow but I get this error then
```
Worker 'prefect-worker-ddp-2' submitting flow run '8379079c-b9e5-4e8c-8fbe-636f60ff77a7'
11:49:31 AM
prefect.flow_runs.worker
Opening process...
11:49:31 AM
prefect.flow_runs.worker
Completed submission of flow run '8379079c-b9e5-4e8c-8fbe-636f60ff77a7'
11:49:31 AM
prefect.flow_runs.worker
Downloading flow code from storage at '/app'
11:49:32 AM
prefect.flow_runs
Created subflow run 'nickel-snake' for flow 'run-airbyte-connection-flow-v1'
11:49:33 AM
prefect.flow_runs
Encountered exception during execution:
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/httpx/_transports/default.py", line 69, in map_httpcore_exceptions
    yield
  File "/usr/local/lib/python3.10/site-packages/httpx/_transports/default.py", line 373, in handle_async_request
    resp = await self._pool.handle_async_request(req)
  File "/usr/local/lib/python3.10/site-packages/httpcore/_async/connection_pool.py", line 216, in handle_async_request
    raise exc from None
  File "/usr/local/lib/python3.10/site-packages/httpcore/_async/connection_pool.py", line 196, in handle_async_request
    response = await connection.handle_async_request(
  File "/usr/local/lib/python3.10/site-packages/httpcore/_async/connection.py", line 99, in handle_async_request
    raise exc
  File "/usr/local/lib/python3.10/site-packages/httpcore/_async/connection.py", line 76, in handle_async_request
    stream = await self._connect(request)
  File "/usr/local/lib/python3.10/site-packages/httpcore/_async/connection.py", line 122, in _connect
    stream = await self._network_backend.connect_tcp(**kwargs)
  File "/usr/local/lib/python3.10/site-packages/httpcore/_backends/auto.py", line 30, in connect_tcp
    return await self._backend.connect_tcp(
  File "/usr/local/lib/python3.10/site-packages/httpcore/_backends/anyio.py", line 114, in connect_tcp
    with map_exceptions(exc_map):
  File "/usr/local/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/usr/local/lib/python3.10/site-packages/httpcore/_exceptions.py", line 14, in map_exceptions
    raise to_exc(exc) from exc
httpcore.ConnectError: All connection attempts failed

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 877, in orchestrate_flow_run
    result = await flow_call.aresult()
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 327, in aresult
    return await asyncio.wrap_future(self.future)
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 352, in _run_sync
    result = self.fn(*self.args, **self.kwargs)
  File "/tmp/tmpru5nmvdsprefect/proxy/prefect_flows.py", line 246, in deployment_schedule_flow_v4
    run_airbyte_connection_flow_v1(task_config)
  File "/usr/local/lib/python3.10/site-packages/prefect/flows.py", line 1246, in __call__
    return enter_flow_run_engine_from_flow_call(
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 293, in enter_flow_run_engine_from_flow_call
    retval = from_sync.wait_for_call_in_loop_thread(
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/api.py", line 218, in wait_for_call_in_loop_thread
    return call.result()
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 318, in result
    return self.future.result(timeout=timeout)
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 179, in result
    return self.__get_result()
  File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 389, in _run_async
    result = await coro
  File "/usr/local/lib/python3.10/site-packages/prefect/client/utilities.py", line 100, in with_injected_client
    return await fn(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 752, in create_and_begin_subflow_run
    return await terminal_state.result(fetch=True)
  File "/usr/local/lib/python3.10/site-packages/prefect/states.py", line 91, in _get_state_result
    raise await get_state_exception(state)
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 877, in orchestrate_flow_run
    result = await flow_call.aresult()
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 327, in aresult
    return await asyncio.wrap_future(self.future)
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 352, in _run_sync
    result = self.fn(*self.args, **self.kwargs)
  File "/tmp/tmpru5nmvdsprefect/proxy/prefect_flows.py", line 48, in run_airbyte_connection_flow_v1
    result = run_connection_sync(connection_block)
  File "/usr/local/lib/python3.10/site-packages/prefect/flows.py", line 1246, in __call__
    return enter_flow_run_engine_from_flow_call(
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 293, in enter_flow_run_engine_from_flow_call
    retval = from_sync.wait_for_call_in_loop_thread(
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/api.py", line 218, in wait_for_call_in_loop_thread
    return call.result()
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 318, in result
    return self.future.result(timeout=timeout)
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 179, in result
    return self.__get_result()
  File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 389, in _run_async
    result = await coro
  File "/usr/local/lib/python3.10/site-packages/prefect/client/utilities.py", line 100, in with_injected_client
    return await fn(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 752, in create_and_begin_subflow_run
    return await terminal_state.result(fetch=True)
  File "/usr/local/lib/python3.10/site-packages/prefect/states.py", line 91, in _get_state_result
    raise await get_state_exception(state)
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 877, in orchestrate_flow_run
    result = await flow_call.aresult()
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 327, in aresult
    return await asyncio.wrap_future(self.future)
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 389, in _run_async
    result = await coro
  File "/usr/local/lib/python3.10/site-packages/prefect_airbyte/flows.py", line 57, in run_connection_sync
    airbyte_sync = await task(airbyte_connection.trigger.aio)(airbyte_connection)
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/api.py", line 150, in wait_for_call_in_loop_thread
    return call.result()
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 318, in result
    return self.future.result(timeout=timeout)
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 179, in result
    return self.__get_result()
  File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 389, in _run_async
    result = await coro
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 1555, in get_task_call_return_value
    return await future._result()
  File "/usr/local/lib/python3.10/site-packages/prefect/futures.py", line 237, in _result
    return await final_state.result(raise_on_failure=raise_on_failure, fetch=True)
  File "/usr/local/lib/python3.10/site-packages/prefect/states.py", line 91, in _get_state_result
    raise await get_state_exception(state)
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 2103, in orchestrate_task_run
    result = await call.aresult()
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 327, in aresult
    return await asyncio.wrap_future(self.future)
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 389, in _run_async
    result = await coro
  File "/usr/local/lib/python3.10/site-packages/prefect_airbyte/connections.py", line 363, in trigger
    async with self.airbyte_server.get_client(
  File "/usr/local/lib/python3.10/site-packages/prefect_airbyte/client.py", line 315, in __aenter__
    await self.check_health_status(self._client)
  File "/usr/local/lib/python3.10/site-packages/prefect_airbyte/client.py", line 73, in check_health_status
    response = await client.get(get_connection_url)
  File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1801, in get
    return await self.request(
  File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1574, in request
    return await self.send(request, auth=auth, follow_redirects=follow_redirects)
  File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1661, in send
    response = await self._send_handling_auth(
  File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1689, in _send_handling_auth
    response = await self._send_handling_redirects(
  File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1726, in _send_handling_redirects
    response = await self._send_single_request(request)
  File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1763, in _send_single_request
    response = await transport.handle_async_request(request)
  File "/usr/local/lib/python3.10/site-packages/httpx/_transports/default.py", line 372, in handle_async_request
    with map_httpcore_exceptions():
  File "/usr/local/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/usr/local/lib/python3.10/site-packages/httpx/_transports/default.py", line 86, in map_httpcore_exceptions
    raise mapped_exc(message) from exc
httpx.ConnectError: All connection attempts failed
11:49:33 AM
prefect.flow_runs
Finished in state Failed('Flow run encountered an exception. ConnectError: All connection attempts failed')
11:49:33 AM
prefect.flow_runs
Process 12 exited cleanly.
```
Looks like its not able to connect to the database ? For this experiment i am using my local postgres database that is also running in a docker container. 

We haven't faced this error `cannot import name 'safe _get_annotation' from 'griffe.agents-nodes'` in our current setup yet, after upgrading it to 2.18.3.

**Comment by Ishankoradia** [comment]
Hi @d-jeph , sorry getting back late on this.
I tried with the remote RDS with no firewall like we discussed. But i am gettting the same error

Its able to fetch the code and submit for flow run but then it loses connection. 

Also i had to comment these lines for it to connect to the remote RDS (in `docker-compose.yml`)
```
volumes:
  prefect:
  postgres:
  flows_volume:
# networks:
#   default:
#     name: prefect-network
```




> Hi @d-jeph , i have tested it out. Thanks for making the changes and ugprading the images with the newer prefect version.
> 
> Just want to reconfirm that this is still true ? (second point from my last comment)
> 
> ```
> I tried running one of them , it gave the same error as above. To make it work i had to go into prefect's database, navigate to table depolyment and update the column path to /app. This is absolutely fine, we could run a script that does this for our current deployments if/when we move to dockerize production deployment.
> ```
> 
> After doing the above, it is able to download the code & submit flow but I get this error then
> 
> ```
> Worker 'prefect-worker-ddp-2' submitting flow run '8379079c-b9e5-4e8c-8fbe-636f60ff77a7'
> 11:49:31 AM
> prefect.flow_runs.worker
> Opening process...
> 11:49:31 AM
> prefect.flow_runs.worker
> Completed submission of flow run '8379079c-b9e5-4e8c-8fbe-636f60ff77a7'
> 11:49:31 AM
> prefect.flow_runs.worker
> Downloading flow code from storage at '/app'
> 11:49:32 AM
> prefect.flow_runs
> Created subflow run 'nickel-snake' for flow 'run-airbyte-connection-flow-v1'
> 11:49:33 AM
> prefect.flow_runs
> Encountered exception during execution:
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.10/site-packages/httpx/_transports/default.py", line 69, in map_httpcore_exceptions
>     yield
>   File "/usr/local/lib/python3.10/site-packages/httpx/_transports/default.py", line 373, in handle_async_request
>     resp = await self._pool.handle_async_request(req)
>   File "/usr/local/lib/python3.10/site-packages/httpcore/_async/connection_pool.py", line 216, in handle_async_request
>     raise exc from None
>   File "/usr/local/lib/python3.10/site-packages/httpcore/_async/connection_pool.py", line 196, in handle_async_request
>     response = await connection.handle_async_request(
>   File "/usr/local/lib/python3.10/site-packages/httpcore/_async/connection.py", line 99, in handle_async_request
>     raise exc
>   File "/usr/local/lib/python3.10/site-packages/httpcore/_async/connection.py", line 76, in handle_async_request
>     stream = await self._connect(request)
>   File "/usr/local/lib/python3.10/site-packages/httpcore/_async/connection.py", line 122, in _connect
>     stream = await self._network_backend.connect_tcp(**kwargs)
>   File "/usr/local/lib/python3.10/site-packages/httpcore/_backends/auto.py", line 30, in connect_tcp
>     return await self._backend.connect_tcp(
>   File "/usr/local/lib/python3.10/site-packages/httpcore/_backends/anyio.py", line 114, in connect_tcp
>     with map_exceptions(exc_map):
>   File "/usr/local/lib/python3.10/contextlib.py", line 153, in __exit__
>     self.gen.throw(typ, value, traceback)
>   File "/usr/local/lib/python3.10/site-packages/httpcore/_exceptions.py", line 14, in map_exceptions
>     raise to_exc(exc) from exc
> httpcore.ConnectError: All connection attempts failed
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 877, in orchestrate_flow_run
>     result = await flow_call.aresult()
>   File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 327, in aresult
>     return await asyncio.wrap_future(self.future)
>   File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 352, in _run_sync
>     result = self.fn(*self.args, **self.kwargs)
>   File "/tmp/tmpru5nmvdsprefect/proxy/prefect_flows.py", line 246, in deployment_schedule_flow_v4
>     run_airbyte_connection_flow_v1(task_config)
>   File "/usr/local/lib/python3.10/site-packages/prefect/flows.py", line 1246, in __call__
>     return enter_flow_run_engine_from_flow_call(
>   File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 293, in enter_flow_run_engine_from_flow_call
>     retval = from_sync.wait_for_call_in_loop_thread(
>   File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/api.py", line 218, in wait_for_call_in_loop_thread
>     return call.result()
>   File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 318, in result
>     return self.future.result(timeout=timeout)
>   File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 179, in result
>     return self.__get_result()
>   File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
>     raise self._exception
>   File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 389, in _run_async
>     result = await coro
>   File "/usr/local/lib/python3.10/site-packages/prefect/client/utilities.py", line 100, in with_injected_client
>     return await fn(*args, **kwargs)
>   File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 752, in create_and_begin_subflow_run
>     return await terminal_state.result(fetch=True)
>   File "/usr/local/lib/python3.10/site-packages/prefect/states.py", line 91, in _get_state_result
>     raise await get_state_exception(state)
>   File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 877, in orchestrate_flow_run
>     result = await flow_call.aresult()
>   File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 327, in aresult
>     return await asyncio.wrap_future(self.future)
>   File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 352, in _run_sync
>     result = self.fn(*self.args, **self.kwargs)
>   File "/tmp/tmpru5nmvdsprefect/proxy/prefect_flows.py", line 48, in run_airbyte_connection_flow_v1
>     result = run_connection_sync(connection_block)
>   File "/usr/local/lib/python3.10/site-packages/prefect/flows.py", line 1246, in __call__
>     return enter_flow_run_engine_from_flow_call(
>   File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 293, in enter_flow_run_engine_from_flow_call
>     retval = from_sync.wait_for_call_in_loop_thread(
>   File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/api.py", line 218, in wait_for_call_in_loop_thread
>     return call.result()
>   File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 318, in result
>     return self.future.result(timeout=timeout)
>   File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 179, in result
>     return self.__get_result()
>   File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
>     raise self._exception
>   File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 389, in _run_async
>     result = await coro
>   File "/usr/local/lib/python3.10/site-packages/prefect/client/utilities.py", line 100, in with_injected_client
>     return await fn(*args, **kwargs)
>   File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 752, in create_and_begin_subflow_run
>     return await terminal_state.result(fetch=True)
>   File "/usr/local/lib/python3.10/site-packages/prefect/states.py", line 91, in _get_state_result
>     raise await get_state_exception(state)
>   File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 877, in orchestrate_flow_run
>     result = await flow_call.aresult()
>   File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 327, in aresult
>     return await asyncio.wrap_future(self.future)
>   File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 389, in _run_async
>     result = await coro
>   File "/usr/local/lib/python3.10/site-packages/prefect_airbyte/flows.py", line 57, in run_connection_sync
>     airbyte_sync = await task(airbyte_connection.trigger.aio)(airbyte_connection)
>   File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/api.py", line 150, in wait_for_call_in_loop_thread
>     return call.result()
>   File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 318, in result
>     return self.future.result(timeout=timeout)
>   File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 179, in result
>     return self.__get_result()
>   File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
>     raise self._exception
>   File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 389, in _run_async
>     result = await coro
>   File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 1555, in get_task_call_return_value
>     return await future._result()
>   File "/usr/local/lib/python3.10/site-packages/prefect/futures.py", line 237, in _result
>     return await final_state.result(raise_on_failure=raise_on_failure, fetch=True)
>   File "/usr/local/lib/python3.10/site-packages/prefect/states.py", line 91, in _get_state_result
>     raise await get_state_exception(state)
>   File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 2103, in orchestrate_task_run
>     result = await call.aresult()
>   File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 327, in aresult
>     return await asyncio.wrap_future(self.future)
>   File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 389, in _run_async
>     result = await coro
>   File "/usr/local/lib/python3.10/site-packages/prefect_airbyte/connections.py", line 363, in trigger
>     async with self.airbyte_server.get_client(
>   File "/usr/local/lib/python3.10/site-packages/prefect_airbyte/client.py", line 315, in __aenter__
>     await self.check_health_status(self._client)
>   File "/usr/local/lib/python3.10/site-packages/prefect_airbyte/client.py", line 73, in check_health_status
>     response = await client.get(get_connection_url)
>   File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1801, in get
>     return await self.request(
>   File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1574, in request
>     return await self.send(request, auth=auth, follow_redirects=follow_redirects)
>   File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1661, in send
>     response = await self._send_handling_auth(
>   File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1689, in _send_handling_auth
>     response = await self._send_handling_redirects(
>   File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1726, in _send_handling_redirects
>     response = await self._send_single_request(request)
>   File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1763, in _send_single_request
>     response = await transport.handle_async_request(request)
>   File "/usr/local/lib/python3.10/site-packages/httpx/_transports/default.py", line 372, in handle_async_request
>     with map_httpcore_exceptions():
>   File "/usr/local/lib/python3.10/contextlib.py", line 153, in __exit__
>     self.gen.throw(typ, value, traceback)
>   File "/usr/local/lib/python3.10/site-packages/httpx/_transports/default.py", line 86, in map_httpcore_exceptions
>     raise mapped_exc(message) from exc
> httpx.ConnectError: All connection attempts failed
> 11:49:33 AM
> prefect.flow_runs
> Finished in state Failed('Flow run encountered an exception. ConnectError: All connection attempts failed')
> 11:49:33 AM
> prefect.flow_runs
> Process 12 exited cleanly.
> ```
> 
> Looks like its not able to connect to the database ? For this experiment i am using my local postgres database that is also running in a docker container.
> 
> We haven't faced this error `cannot import name 'safe _get_annotation' from 'griffe.agents-nodes'` in our current setup yet, after upgrading it to 2.18.3.

**Review by fatchat — APPROVED** [review]

**Review by Ishankoradia — APPROVED** [review]

**Pull request merged**